Use `ismissing(x)` instead of `x === missing`

nalimilan opened the pull request:
This is more generic and will allow packages to define custom `missing`-like types allowing to distinguish several kinds of
missing values like in e.g. Stata and SAS (see https://github.com/nalimilan/TypedMissings.jl). This should have no performance impact now thanks to #38905.

https://github.com/JuliaCI/BaseBenchmarks.jl/pull/294 adds more benchmarks to cover changes made in this PR, let's see whether Nanosoldier confirms that no performance regressions happen.

nalimilan committed:
Use `ismissing(x)` instead of `x === missing`

This is more generic and will allow packages to define custom
`missing`-like types allowing to distinguish several kinds of
missing values like in e.g. Stata and SAS. This should have
no performance impact now thanks to #38905 and #49301.

nalimilan commented:
@nanosoldier `runbenchmarks("union", vs = ":master")`

KristofferC commented:
>  This should have no performance impact

I guess that depends on what you mean, for example:

```jl
julia> f_is(x) = ismissing(x);

julia> f_egal(x) = x === missing;

julia> Core.Compiler.return_type(f_is, Tuple{Any})
Bool

julia> Core.Compiler.return_type(f_egal, Tuple{Any})
Bool

julia> struct A end; struct B end; struct C end; struct D end;

julia> Base.ismissing(::A) = true; Base.ismissing(::B) = true; Base.ismissing(::C) = true; Base.ismissing(::D) = true

julia> Core.Compiler.return_type(f_is, Tuple{Any})
Any

julia> Core.Compiler.return_type(f_egal, Tuple{Any})
Bool
```

nalimilan commented:
Yeah there will always be cases where `===` is more efficient than a function call. Let's say that I hope this won't have performance impacts in standard use cases where performance is currently good. Overloading `ismissing` for custom types doesn't count, as that's precisely something that is not really supported currently: without this PR, it's ignored by essential functions dealing with missing values. That said, couldn't the compiler handle your example better, given that all defined functions return a `Bool`? I guess it just doesn't consider that computing the return type is worth the compilation cost?

BTW, any idea why Nanosoldier doesn't seem to react?

vtjnash commented:
You have hidden your access permission status (https://github.com/orgs/JuliaLang/people/nalimilan), so nanosoldier cannot authenticate you

vtjnash commented on base/missing.jl:
```diff
@@ -459,7 +459,7 @@ ERROR: `b` is still missing
     expr = :(missing)
     for arg in reverse(args)
-        expr = :((val = $arg) !== missing ? val : $expr)
+        expr = :(!ismissing((val = $arg;)) ? val : $expr)
     end
     return esc(:(let val; $expr; end))
```
This seems to have hygiene problems (calls the wrong `ismissing` function). I think this would be correct:
```suggestion
        expr = :(!ismissing((val = $(esc(arg));)) ? val : $expr)
    end
    return :(local val; $expr; end)
```

nalimilan replied:
Thanks. I'm very bad at macros programming... Though your suggestion didn't work as-is, I went back to using `let`. Was `local` intended or just a typo?

nalimilan committed:
Fix macro hygiene

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

nalimilan committed:
Fix the fix

nalimilan commented:
@nanosoldier runbenchmarks("union", vs = ":master")

nalimilan commented:
> You have hidden your access permission status (https://github.com/orgs/JuliaLang/people/nalimilan), so nanosoldier cannot authenticate you

Indeed, thanks. Though I've just changed my status to Public, but Nanosoldier doesn't want to start. The link above still doesn't work, but maybe that's because I'm not an owner?

fredrikekre commented:
You need to use backticks as in your original attempt.

nalimilan commented:
Shame on me! I swear I used to know how to do it. :-D
`@nanosoldier runbenchmarks("union", vs = ":master")`

nalimilan commented:
@nanosoldier `runbenchmarks("union", vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/44407#issuecomment-1059048016) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/c0c82a7_vs_b63ae3b/report.md).

nalimilan commented:
OK, there are many small improvements and regressions which appear to be random noise. The only worrying one is the 3.73 time increase for `["union", "array", ("skipmissing", "sum", "Union{Missing, Int8}", 1)]`, but looking at the generated code locally:
```julia
@code_llvm Base._mapreduce(identity, Base.add_sum, IndexLinear(), skipmissing(Union{Int8, Missing})[])
```
it's identical to master, so AFAICT that's random too.

Let's run it again just to be sure:
@nanosoldier `runbenchmarks("union", vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/44407#issuecomment-1059731751) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/e40209a_vs_2349f0a/report.md).

nalimilan commented:
Hmm, the regression is still there. Looking at `@code_llvm` for `mapreduce_impl`, this is due to SIMD not being used. A MWE is:
```julia
A = Vector{Union{Int8, Missing}}(rand(Int8, 10_000));
A[rand(10_000) .> .9] .= missing;
@btime Base.mapreduce_impl(identity, Base.add_sum, skipmissing(A), 1, length(A), 1024)
```
Now what's funny is that SIMD is used for `Int64`, `Int32` and `Int16`, the only exception is `Int8` (for which the performance gain would be the largest). So in the end summing `Int16` values is 5 times faster than summing `Int8` values on this PR while there's no difference on master. Any ideas what may be causing this?

nalimilan commented:
The performance regression appears to have been fixed on master since 1.10.0-rc1. Let's confirm this with Nanosoldier.

@nanosoldier `runbenchmarks("union", vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/44407#issuecomment-1829265984) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/8cd4e34_vs_72dbbdb/report.md).

nalimilan commented:
Regression is indeed gone. I'll merge in a few days if nobody objects.

mkitti commented:
I have a duplicate in #52341 .

One thing that may be worth grabbing is a test for a new type `NewMissing` that has `ismissing(::NewMissing) = true`
https://github.com/JuliaLang/julia/pull/52341/commits/971a173b062a37c0d514db021214b0f7bd648cd0

That permits `skipmissing` to work at least.

nalimilan committed:
Add tests, drop use of >: Missing

nalimilan commented:
Good idea. I've copied your tests and added more of them. This uncovered another place where `>: Missing` was hardcoded and triggered an error in reductions. I've dropped that check as my benchmarks didn't show any performance difference (which makes sense since finding the first non-missing values in a vector whose type doesn't support missing values is super fast).

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/44407#issuecomment-1837104976) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/f15ff98_vs_aef528d/report.md).

nalimilan commented:
Benchmarks seem to indicate that PR makes `sum(skipmissing(::Vector{<:Real}))` slower by 6-21%. Not sure whether it matters a lot given that `skipmissing` isn't really supposed to be used with vectors that do not allow for missing values. It would be easy to keep the existing fast path for `AbstractVector{<:Real}`, but that wouldn't cover other types.

mkitti commented:
I would focus on the 21% case.

```
["union", "array", ("skipmissing", "sum", "Float32", 0)]
```

nalimilan commented:
Well, the fix is the same for all types. The problem is just that a generic fix that would cover types that are not in the benchmarks (e.g. `Date` or any type defined in a package) isn't really possible (except for reverting the change).

nalimilan committed:
Revert `>: Missing` change

nalimilan committed:
Mark test as broken

nalimilan commented:
OK, I've reverted that change for now as it's more problematic than using `ismissing`, which was the original goal of this PR. I'll merge in a few days if nobody objects, as only tests have changed since @vtjnash approved last year (!).
